### ISSUE TITLE
Last decoder output feature should be 8 instead is 32

### ISSUE THREAD

**Issue by lsaeuro** [qiqihaer/RandLA-Net-pytorch#20]
Hi, thank you to your work, it is very useful for me. 
BTW, looking at the model structure I noticed a difference with the original network, in the last decoder dimension. Your structure is :  
`(decoder_blocks): ModuleList(
    (0): Conv2d(
      (conv): Conv2d(768, 256, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn): BatchNorm2d(
        (bn): BatchNorm2d(256, eps=1e-06, momentum=0.99, affine=True, track_running_stats=True)
      )
      (activation): LeakyReLU(negative_slope=0.2, inplace=True)
    )
    (1): Conv2d(
      (conv): Conv2d(384, 128, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn): BatchNorm2d(
        (bn): BatchNorm2d(128, eps=1e-06, momentum=0.99, affine=True, track_running_stats=True)
      )
      (activation): LeakyReLU(negative_slope=0.2, inplace=True)
    )
    (2): Conv2d(
      (conv): Conv2d(160, 32, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn): BatchNorm2d(
        (bn): BatchNorm2d(32, eps=1e-06, momentum=0.99, affine=True, track_running_stats=True)
      )
      (activation): LeakyReLU(negative_slope=0.2, inplace=True)
    )
    (3): Conv2d(
      (conv): Conv2d(64, 32, kernel_size=(1, 1), stride=(1, 1), bias=False)
      (bn): BatchNorm2d(
        (bn): BatchNorm2d(32, eps=1e-06, momentum=0.99, affine=True, track_running_stats=True)
      )
      (activation): LeakyReLU(negative_slope=0.2, inplace=True)
    )
  )`

We can see that the last layer of the decoder has 32 feature while according to the network it should be downsampled by 4, so it should be 8. 
![image](https://user-images.githubusercontent.com/107695630/176623574-567a3b41-0a74-45a3-b24f-25b98fad681f.png)

Moreover we can see it also here, where for an input point cloud of 1000 points ( a sample pc), with batch_size = 6, we obtain as last decoder layer torch.Size([6, 32, 1000, 1]). 

My suggestion is to write [here](https://github.com/qiqihaer/RandLA-Net-pytorch/blob/a255652b65d8378682479c23299704783a8fe4d9/RandLANet.py#L36) , d_out = 8, or something less hard coded, but with 8 as dimension.

If I am wrong, could you please explain me the meaning and the reason of that output?
Thank you in advance

**Comment by 773041642** [comment]
The output of the official code is 32 dimensions.

**Comment by lsaeuro** [comment]
> The output of the official code is 32 dimensions.

Yes indeed i am talking about the decoder output, not the network output

**Comment by lsaeuro** [comment]
> > The output of the official code is 32 dimensions.
> 
> Yes indeed i am talking about the decoder output, not the network output

Anyway, is it or not N=8 in the paper? I am confused on why they changed it

**Comment by 773041642** [comment]
> > > The output of the official code is 32 dimensions.
> > 
> > 
> > Yes indeed i am talking about the decoder output, not the network output
> 
> Anyway, is it or not N=8 in the paper? I am confused on why they changed it

I know what you mean, the decoder output of the official code is 32 dimensions, you can debug it to look the result.

![图片](https://user-images.githubusercontent.com/56188697/177548929-3b847ab2-bbed-4fd1-8326-6a3f558c7531.png)

**Comment by 773041642** [comment]
> > > > The output of the official code is 32 dimensions.
> > > 
> > > 
> > > Yes indeed i am talking about the decoder output, not the network output
> > 
> > 
> > Anyway, is it or not N=8 in the paper? I am confused on why they changed it
> 
> I know what you mean, the decoder output of the official code is 32 dimensions, you can debug it to look the result.

PICTURE:/home/lxx/.config/tencent-qq//AppData/file//sendpix0.jpg

**Comment by lsaeuro** [comment]
Thank you, I see.. I will try both implementation to see which one performs better at this point I think

**Comment by 773041642** [comment]
> Thank you, I see.. I will try both implementation to see which one performs better at this point I think

Do you know the reason for this processing? I haven't found any disordered operations in the official code.

![image](https://user-images.githubusercontent.com/56188697/177575461-964f780c-0a51-4831-81a6-6d2d91562a1b.png)

Thank you very much!

**Comment by lsaeuro** [comment]
I think I do not have the answer  to this... Do you know where  in the paper is mentioned this probability sampling strategy? @773041642

**Comment by 773041642** [comment]
Random sampling is not specifically explained in the paper. You can take a look at the implementation method in the code, or see other people's explanations about the paper.

**Comment by lsaeuro** [comment]
Thank you for your answers..
I was looking at the code to figure out also if they use the variable 'cloud_ind' , in this Pytorch Version it is never used while in the TF version they use it here 
![image](https://user-images.githubusercontent.com/107695630/177762698-3cb98ec6-12e9-4aaf-93f4-1c69bc9fbc11.png)

Do you know the reason why?

**Comment by huixiancheng** [comment]
It's only use for test full point cloud. 
This repo don't generate test. So it's unuseful.
